### ISSUE TITLE
Soft-forkability

### ISSUE THREAD

**Issue by catena2w** [ergoplatform/ergo#581]
The goal of this issue is to allow to update protocol via soft forks after mainnet launch.

**Comment by catena2w** [comment]
Proposal: 
- Numerate all the validation rules, like
```
    failFast
      .demand(1, outputs.forall(o => o.value >= BoxUtils.minimalErgoAmount(o, stateContext.currentParameters)))
      .demand(2, outputCandidates.forall(_.creationHeight <= stateContext.currentHeight))
     ...
    .demand(N, isCostValid)
```
- When a soft-fork is proposed, a set of validation rules that are chaged should be described in extension. E.g. if we're proposing a soft-fork, that changes costing rules and minimal output value, it should specify `[1,N]` as chagned validation rules
- When a soft fork is activated, while some node is still on a previous version of the protocol, this node skips specified checks.

Notes:
- PoW validation should not be changed via soft-fork.
- This will enforce us to unify our checks via `ModifierValidator`, that is a good and long awaited refactoring
- While doing this task, I propose to create a table with all validation rules in yellow paper, and describe validation of all modifiers as just a list of validation rules. E.g. in such a way ErgoTransaction validation will be described as `checks [1,2,...,N]`
- There might be higher and lower level validations, e.g. soft-fork may disable the full `BlockTransactions` section validation (for sure it's a dirty way of doing soft-fork)
- If this will be implemented, I propose to implement https://github.com/ergoplatform/ergo/issues/580 via soft-fork after mainnet launch

**Comment by catena2w** [comment]
fixed in #754
